README.md
### ms

a utility for converting time descriptions (i.e "2hrs 30mins") into milliseconds.

### Installation

```
npm install @ededejr/ms
```

### Usage

```ts
import { ms } from "@ededejr/ms";

ms("1 min"); // 60000
```

##### What can I parse?

###### Hours and minutes

```ts
ms("2hrs"); // 7200000
ms("2hrs 30mins"); // 9000000;
```

###### Days

```ts
ms("10 days"); // 864000000
ms("1 day 20hrs 40mins"); // 160800000
```

###### AM/PM Times

```ts
ms("10AM"); // ms from now till the next time it will be 10AM
ms("10:30PM"); // ms from now till the next time it will be 10:30PM
```


CHANGELOG.md


# [0.2.0](https://github.com/ededejr/ms/compare/v0.1.0...v0.2.0) (2023-05-26)


### Bug Fixes

* update comment for SecsRGX ([cb38ef1](https://github.com/ededejr/ms/commit/cb38ef1a4de880dfd53f4c46a220f1484a61236d))
* update param in function comment ([0a624c1](https://github.com/ededejr/ms/commit/0a624c1f5fd9c9d0138ce0155ba03ee931ded350))


### Features

* complete transition to StandardRGX ([cb7c9f3](https://github.com/ededejr/ms/commit/cb7c9f352eced7c739b0e5871c192666aa218a00))
* return NaN instead of undefined ([7094f53](https://github.com/ededejr/ms/commit/7094f53733c3054849631c36423c7ae0214983ad))
* StandardRGX combining all other matchers ([800918c](https://github.com/ededejr/ms/commit/800918cd9274d26105e0942532cb0089a2379664))

# [0.1.0](https://github.com/ededejr/ms/compare/v0.0.4...v0.1.0) (2023-05-26)


### Bug Fixes

* import issues after deletion ([3c40806](https://github.com/ededejr/ms/commit/3c40806a177ea6ffe0d179a9f568ca7511c8943b))


### Features

* support template literals ([c1c6610](https://github.com/ededejr/ms/commit/c1c6610fda1f9848c4e49a8955ae45d02aa5f0b0))

## [0.0.4](https://github.com/ededejr/ms/compare/v0.0.3...v0.0.4) (2021-12-10)

## 0.0.3 (2020-10-07)

src/tests/index.test.ts
import { ms } from '../index';

function createBasicTest(input: string, expected: number | undefined, label?: string) {
  test(label ? `${label}: "${input}"` : `"${input}"`, () => expect(ms(input)).toBe(expected));
}

describe("seconds", () => {
  createBasicTest('1 sec', 1000);
  createBasicTest('1sec', 1000);
  createBasicTest('1 secs', 1000);
  createBasicTest('1secs', 1000);
  createBasicTest('1 second', 1000);
  createBasicTest('1second', 1000);
  createBasicTest('1 seconds', 1000);
  createBasicTest('1seconds', 1000);
  createBasicTest('20sec', 20000);
  createBasicTest('20 secs', 20000);
  createBasicTest('20 second', 20000);
  createBasicTest('20 seconds', 20000);
  createBasicTest('100 seconds', 100000);
  createBasicTest('300 seconds', 300000);
  createBasicTest('300s', 300000);
  createBasicTest('1s', 1000);
  createBasicTest('1 s', 1000);
  createBasicTest('1000s', 1000000);
});

describe("minutes", () => {
  createBasicTest('1 min', 60000);
  createBasicTest('1min', 60000);
  createBasicTest('1 mins', 60000);
  createBasicTest('1mins', 60000);
  createBasicTest('1 minute', 60000);
  createBasicTest('1minute', 60000);
  createBasicTest('1 minutes', 60000);
  createBasicTest('1minutes', 60000);
  createBasicTest('20min', 1200000);
  createBasicTest('20 mins', 1200000);
  createBasicTest('20 minute', 1200000);
  createBasicTest('20 minutes', 1200000);
  createBasicTest('100 minutes', 6000000);
  createBasicTest('300 minutes', 18000000);
});

describe("hours and minutes", () => {
  createBasicTest('2hrs', 7200000);
  createBasicTest('2hrs 30mins', 9000000);
});

describe("days", () => {
  createBasicTest('1 day', 86400000);
  createBasicTest('1 days', 86400000);
  createBasicTest('1day', 86400000);
  createBasicTest('1days', 86400000);
  createBasicTest('1 day 20hrs 40mins', 160800000);
  createBasicTest('1 day 25hrs', 176400000);
  createBasicTest('10 day', 864000000);
})

describe("validity", () => {
  test('should throw an error, when ms(undefined)', () => {
    expect(() => ms(undefined)).toThrowError();
  });

  test('should return NaN when ms("☃")', () => {
    expect(ms('☃')).toBeNaN();
  });

  test('should return NaN with decimals', () => {
    expect(ms('1.5')).toBeNaN();
  });

  test('should return NaN with decimals and units', () => {
    expect(ms('1.5s')).toBeNaN();
  });
})


src/tests/regex.test.ts
import { AmPmRGX, StandardRGX } from "../regexes"

describe('StandardRGX', () => {
  function makeTest(input: string, expected: boolean) {
    test(`should ${!expected ? "not " : ''}match "${input}"`, () => {
      expect(StandardRGX.matches(input)).toBe(expected);
    });
  }

  describe('Valid time expressions', () => {
    makeTest("2 years 3 weeks 4 days 5 hours 6 minutes 7 seconds 8 milliseconds", true)
    makeTest("1 year 2 weeks 3 days", true)
    makeTest("6 hours 30 minutes", true)
    makeTest("15 seconds", true)
    makeTest("500 milliseconds", true)
  });

  describe('Valid time expressions with optional components omitted', () => {
    makeTest("1 year 6 minutes", true)
    makeTest("2 days 10 hours", true)
    makeTest("20 minutes 300 milliseconds", true)
  });

  describe('Individual components', () => {
    makeTest("2 hours", true)
    makeTest("3 weeks", true)
  });

  describe('Individual components with plural and singular units', () => {
    makeTest("1 year", true)
    makeTest("2 minutes", true)
    makeTest("3 seconds", true)
    makeTest("1 day", true)
  });

  describe('Invalid input', () => {
    makeTest("1", false);
    makeTest("1 year 2 weeks 3 hours 4 invalidunit", false)
    makeTest("1 hour 2 invalidunits", false)
    makeTest("random string", false)
  });

  describe('Time expressions with spaces', () => {
    makeTest("2years 3weeks 4days 5hours 6minutes 7seconds 8milliseconds", true)
    makeTest("1year 2weeks 3days", true)
  });

  describe('Time expressions with mixed case', () => {
    makeTest("1 Year 2 wEeKs 3 dAYS 4 hOUrS 5 mINutEs 6 SECondS 7 MiLliSecoNds", true)
  });

  describe('Time expressions with leading/trailing spaces', () => {
    makeTest(" 5 hours 10 minutes ", true)
    makeTest(" 1 year ", true)
  })

  describe('Time expressions with zero values', () => {
    makeTest("0 years 0 weeks 0 days 0 hours 0 minutes 0 seconds 0 milliseconds", true)
  })
})

describe("AmPmRGX", () => {
  function makeTest(input: string, expected: boolean) {
    test(`should ${!expected ? "not " : ''}match "${input}"`, () => {
      expect(AmPmRGX.matches(input)).toBe(expected);
    });
  }

  makeTest("10AM", true)
  makeTest("10:30AM", true)
  makeTest("10:30 AM", true)
  
  makeTest("10:30PM", true)
  makeTest("10:30 PM", true)
  makeTest("10PM", true)

  makeTest("10:30am", true)
  makeTest("10:30 am", true)
  makeTest("10am", true)

  makeTest("10:30pm", true)
  makeTest("10:30 pm", true)
  makeTest("10pm", true)

  makeTest("00:00pm", true);

  makeTest("10:30", false)
  makeTest("10", false)
  makeTest("10:30:30", false)
  makeTest("10:30:30AM", false)
  makeTest("10:30:30PM", false)
});

src/tests/vercel.test.ts
/**
 * Taken and adapted from @vercel/ms
 * @see https://github.com/vercel/ms/blob/master/src/index.test.ts
 */

import { ms } from '../';

describe('ms(string)', () => {
  it('should not throw an error', () => {
    expect(() => {
      ms('1m');
    }).not.toThrowError();
  });

  it('should preserve ms', () => {
    expect(ms('100')).toBe(100);
  });

  it('should convert from m to ms', () => {
    expect(ms('1m')).toBe(60000);
  });

  it('should convert from h to ms', () => {
    expect(ms('1h')).toBe(3600000);
  });

  it('should convert d to ms', () => {
    expect(ms('2d')).toBe(172800000);
  });

  it('should convert w to ms', () => {
    expect(ms('3w')).toBe(1814400000);
  });

  it('should convert s to ms', () => {
    expect(ms('1s')).toBe(1000);
  });

  it('should convert ms to ms', () => {
    expect(ms('100ms')).toBe(100);
  });

  it('should convert y to ms', () => {
    expect(ms('1y')).toBe(31536000000);
  });

  it('should work with multiple spaces', () => {
    expect(ms('1   s')).toBe(1000);
  });

  it('should work with negative integers', () => {
    expect(ms('-100ms')).toBe(-100);
  });
});

// long strings

describe('ms(long string)', () => {
  it('should not throw an error', () => {
    expect(() => {
      ms('53 milliseconds');
    }).not.toThrowError();
  });

  it('should convert milliseconds to ms', () => {
    expect(ms('53 milliseconds')).toBe(53);
  });

  it('should convert sec to ms', () => {
    expect(ms('1 sec')).toBe(1000);
  });

  it('should convert from min to ms', () => {
    expect(ms('1 min')).toBe(60000);
  });

  it('should convert from hr to ms', () => {
    expect(ms('1 hr')).toBe(3600000);
  });

  it('should convert days to ms', () => {
    expect(ms('2 days')).toBe(172800000);
  });

  it('should convert weeks to ms', () => {
    expect(ms('1 week')).toBe(604800000);
  });

  it('should convert years to ms', () => {
    expect(ms('1 year')).toBe(31536000000);
  });

  it('should work with negative integers', () => {
    expect(ms('-100 milliseconds')).toBe(-100);
  });
});

// invalid inputs

describe('ms(invalid inputs)', () => {
  it('should throw an error, when ms("")', () => {
    expect(() => {
      ms('');
    }).toThrowError();
  });

  it('should return NaN, when ms("☃")', () => {
    expect(ms('☃')).toBeNaN();
  });

  it('should throw an error, when ms(undefined)', () => {
    expect(() => {
      ms(undefined);
    }).toThrowError();
  });

  it('should throw an error, when ms(null)', () => {
    expect(() => {
      ms(null);
    }).toThrowError();
  });

  it('should throw an error, when ms([])', () => {
    expect(() => {
      // @ts-expect-error - We expect this to throw.
      ms([]);
    }).toThrowError();
  });

  it('should throw an error, when ms({})', () => {
    expect(() => {
      // @ts-expect-error - We expect this to throw.
      ms({});
    }).toThrowError();
  });

  it('should throw an error, when ms(NaN)', () => {
    expect(() => {
      // @ts-expect-error - We expect this to throw.
      ms(NaN);
    }).toThrowError();
  });

  it('should throw an error, when ms(Infinity)', () => {
    expect(() => {
      // @ts-expect-error - We expect this to throw.
      ms(Infinity);
    }).toThrowError();
  });

  it('should throw an error, when ms(-Infinity)', () => {
    expect(() => {
      // @ts-expect-error - We expect this to throw.
      ms(-Infinity);
    }).toThrowError();
  });
});

src/regexes.ts
import TimeResolvers, { TimeDelineation } from "./resolvers";

class TimeStringRegex {
  private rgx: RegExp;
  private resolvers: TimeDelineation[];

  constructor(rgx: RegExp, resolvers: TimeDelineation[]) {
    this.rgx = rgx;
    this.resolvers = resolvers;
  }

  /**
   * Get the Javascript `RegexExp` this wraps.
   */
  get regex() {
    return this.rgx;
  }

  /**
   * Check if a time string passes regex.test.
   *
   * @param time The time string to test.
   */
  matches(time: string) {
    time = time.trim().replace(/ /g, "");
    return this.rgx.test(time);
  }

  /**
   * Convert a time string to milliseconds.
   *
   * @param time The time string to convert.
   *
   * @example
   * // this.rgx = /^(?<mins>\d{1,3})min(?:ute)?s?$/;
   *
   * this.convertStringToMilliseconds('30mins') => 1800000
   */
  convertStringToMilliseconds(time: string) {
    if (!this.matches(time)) {
      return;
    }

    const matchResult = time.match(this.rgx);
    const { groups: matchGroups } = matchResult || {};

    if (matchGroups) {
      let totalInMs = 0;

      for (const groupName of Object.keys(matchGroups)) {
        const resolver = this.resolvers.find((g) => g.name === groupName);
        const value = matchGroups[groupName];

        if (value && resolver) {
          const result = resolver.resolve(+value, value, matchGroups) || 0;

          if (resolver.name === "indicator") {
            return result;
          }

          totalInMs += result;
        }
      }

      return totalInMs;
    }
  }
}

/**
 * Checks if a time string is in the format 10AM/10:30AM or 10PM/10:30PM.
 * Lowercase is okay for the AM/PM.
 *
 * @param time The time string to check.
 */
export const AmPmRGX = new TimeStringRegex(
  /^(?<hrs>[0-1]?[0-9])(?:\:(?<mins>[0-5][0-9]))? *(?<indicator>AM|am|PM|pm)$/,
  [TimeResolvers.hrs, TimeResolvers.mins, TimeResolvers.indicator]
);

/**
 * Parse and extract years, weeks, days, hours, minutes, seconds,
 * and milliseconds from a time string.
 * 
 * @param time The time string to check. 
 */
export const StandardRGX = new TimeStringRegex(
  /^(?:(?<yrs>\d+)\s*y(?:ears?)?)?\s*(?:(?<wks>\d+)\s*w(?:eeks?|ks)?)?\s*(?:(?<days>\d+)\s*d(?:ays?)?)?\s*(?:(?<hrs>\d+)\s*h(?:(r|our)s?)?)?\s*(?:(?<mins>\d+)\s*m(?:in(?:ute)?s?)?)?\s*(?:(?<secs>\d+)\s*s(?:ec(?:ond)?s?)?)?\s*(?:(?<ms>\d+)(?:\s*(ms|(?:(mill(?:i)?second)s?))))?$/i,
  [
    TimeResolvers.yrs,
    TimeResolvers.wks,
    TimeResolvers.days,
    TimeResolvers.hrs,
    TimeResolvers.mins,
    TimeResolvers.secs,
    TimeResolvers.ms,
  ]
);


src/resolvers.ts
const ONE_SEC = 1000;
const ONE_MIN = 60 * ONE_SEC;
const ONE_HOUR = 60 * ONE_MIN;
const ONE_DAY = 24 * ONE_HOUR;
const ONE_WEEK = 7 * ONE_DAY;
const ONE_YEAR = ONE_DAY * 365;

export interface TimeDelineation {
  name: string;
  resolve: (
    value: number,
    valueString?: string,
    groups?: { [key: string]: string }
  ) => number;
}

export default {
  ms: {
    name: "ms",
    resolve: (value) => value,
  },
  secs: {
    name: "secs",
    resolve: (value) => value * ONE_SEC,
  },
  mins: {
    name: "mins",
    resolve: (value) => value * ONE_MIN,
  },
  hrs: {
    name: "hrs",
    resolve: (value) => value * ONE_HOUR,
  },
  days: {
    name: "days",
    resolve: (value) => value * ONE_DAY,
  },
  wks: {
    name: "wks",
    resolve: (value) => value * ONE_WEEK,
  },
  yrs: {
    name: "yrs",
    resolve: (value) => value * ONE_YEAR,
  },
  indicator: {
    name: "indicator",
    resolve: (_, valueString, matchGroups) => {
      if (!valueString || !matchGroups) {
        throw new Error(
          "Cannot resolve indicator value without valueString and match groups"
        );
      }

      let { hrs } = matchGroups;
      const { mins } = matchGroups;

      hrs = /am/i.test(valueString) ? hrs : hrs + 12;

      // The parsed date will be incorrect,
      // We just need the date object to do some
      // maths.
      const parsedHour = +hrs;
      const parsedMinutes = +mins || 0;
      const now = new Date();

      const currentHour = now.getHours();
      const currentMinute = now.getMinutes();

      function getTimeTillTomorrow() {
        const minsToMidnight = (60 - currentMinute) * ONE_MIN;
        const hoursTillMidnight =
          (24 - (minsToMidnight === 0 ? currentHour : currentHour + 1)) *
          ONE_HOUR;
        const totalTimeToday = hoursTillMidnight + minsToMidnight;

        const hoursTillTomorrow = parsedHour * ONE_HOUR;
        const minutesTillTomorrow = parsedMinutes * ONE_MIN;
        const totalTimeTomorrow = hoursTillTomorrow + minutesTillTomorrow;

        return totalTimeToday + totalTimeTomorrow;
      }

      // If its greater than now calculate time till then
      // otherwise add the remaining hours in today,
      // plus how many hours we need for tomorrow
      // plus the minutes if any
      // If its morning, check if the time is past
      if (currentHour < parsedHour) {
        const hourDiff = (parsedHour - currentHour) * ONE_HOUR;
        return (
          currentHour +
          hourDiff +
          parsedMinutes * ONE_MIN -
          currentMinute * ONE_MIN
        );
      } else if (currentHour === parsedHour) {
        if (parsedMinutes > currentMinute) {
          return (parsedMinutes - currentMinute) * ONE_MIN;
        } else {
          return getTimeTillTomorrow() - currentMinute * ONE_MIN;
        }
      } else {
        return getTimeTillTomorrow();
      }
    },
  },
} as { [key: string]: TimeDelineation };


scripts/runner.ts
class ScriptLogger {
  constructor(private name: string) {}

  private format(str: string) {
    return `[${this.name}] ${str}`;
  }

  log(message: string) {
    console.log(this.format(message));
  }

  error(message: string) {
    console.error(this.format(message));
  }

  warn(message: string) {
    console.warn(this.format(message));
  }

  table(data: any) {
    console.table(data);
  }
}

interface Services {
  logger: ScriptLogger;
}

export interface Script {
  name: string;
  run: (services: Services) => Promise<void>;
}

export async function runScript(script: Script) {
  console.log(`[runner] Running script: ${script.name}`);

  const logger = new ScriptLogger(script.name);
  try {
    await script.run({ logger });
  } catch (e) {
    logger.error(`Error: ${e.message}`);
    logger.error(e.stack);
  }

  console.log(`[runner] Finished script: ${script.name}`);
}


src/time-resolvers.ts
import { ONE_DAY, ONE_HOUR, ONE_MIN } from "@nend/utils/date";

export interface TimeDelineation {
  name: string,
  resolve: (value: number, valueString?: string, groups?: {[key:string]: string}) => number
}

export default {
  mins: {
    name: 'mins',
    resolve: (value) => value * ONE_MIN
  },
  hrs: {
    name: 'hrs',
    resolve: (value) => value * ONE_HOUR
  },
  days: {
    name: 'days',
    resolve: (value) => value * ONE_DAY
  },
  indicator: {
    name: 'indicator',
    resolve: (_, valueString, matchGroups) => {
      if (!valueString || !matchGroups) {
        throw new Error('Cannot resolve indicator value without valueString and match groups');
      }

      let { hrs } = matchGroups;
      const { mins } = matchGroups;

      hrs = /am/i.test(valueString) ? hrs : hrs + 12;

      // The parsed date will be incorrect,
      // We just need the date object to do some
      // maths.
      const parsedHour = +hrs;
      const parsedMinutes = +mins || 0;
      const now = new Date();

      const currentHour = now.getHours();
      const currentMinute = now.getMinutes();

      function getTimeTillTomorrow () {
        const minsToMidnight = (60 - currentMinute) * ONE_MIN;
        const hoursTillMidnight = (24 - (minsToMidnight === 0 ? currentHour : currentHour + 1)) * ONE_HOUR;
        const totalTimeToday = hoursTillMidnight + minsToMidnight;

        const hoursTillTomorrow = (parsedHour * ONE_HOUR);
        const minutesTillTomorrow = (parsedMinutes * ONE_MIN);
        const totalTimeTomorrow = hoursTillTomorrow + minutesTillTomorrow;

        return totalTimeToday + totalTimeTomorrow;
      }

      // If its greater than now calculate time till then
      // otherwise add the remaining hours in today,
      // plus how many hours we need for tomorrow
      // plus the minutes if any
      // If its morning, check if the time is past
      if (currentHour < parsedHour) {
        const hourDiff = (parsedHour - currentHour) * ONE_HOUR;
        return (currentHour + hourDiff + (parsedMinutes * ONE_MIN)) - (currentMinute * ONE_MIN);
      } else if (currentHour === parsedHour) {
        if (parsedMinutes > currentMinute) {
          return (parsedMinutes - currentMinute) * ONE_MIN;
        } else {
          return getTimeTillTomorrow() - (currentMinute * ONE_MIN);
        }
      } else {
        return getTimeTillTomorrow();
      }
    }
  }
} as {[key:string]: TimeDelineation}

src/index.ts
import { AmPmRGX, StandardRGX } from "./regexes";

/**
 * Converts a time descriptor to milliseconds.
 *
 * @param str The string to convert.
 * 
 * @example
 * ms("30") => 30
 * ms("30mins") => 1800000 
 * ms("2hrs 30mins") => 9000000
 * ms("2hrs 30mins 30secs") => 9030000
 * ms("2hrs 30mins 30secs 30ms") => 9030030
 * ms("2y 2w 2d 2h 2m 2s 2ms") => 64461722002
 */
export function ms(str: TemplateStringsArray | string): number | undefined {
  const isTemplateString =
    Array.isArray(str) && str.length === 1 && typeof str[0] === "string";
  const isString = typeof str === "string";

  if (!(isTemplateString || isString)) {
    throw new Error(errors.invalidInput);
  }

  let input = "";

  if (isString) {
    input = str;
  } else if (isTemplateString) {
    input = str[0];
  }

  input = input.trim().replace(/ /g, "");

  if (input.length === 0) {
    throw new Error(errors.emptyString);
  }

  const isNegative = input.startsWith("-");

  if (isNegative) {
    input = input.slice(1);
  }

  const handleNegative = (value: number) => (isNegative ? -value : value);

  if (AmPmRGX.matches(input)) {
    return handleNegative(AmPmRGX.convertStringToMilliseconds(input));
  }

  if (StandardRGX.matches(input)) {
    return handleNegative(StandardRGX.convertStringToMilliseconds(input));
  }

  try {
    const int = parseInt(input, 10);
    
    if (int.toString() !== input) {
      throw new Error(errors.decimals);
    }

    return handleNegative(int);
  } catch (error) {
   // do nothing 
  }

  return NaN;
}

const errors = {
  invalidInput: "Invalid input. Expected a string or template string.",
  emptyString: "Invalid input. Cannot parse an empty string.",
  decimals: "Invalid input. Cannot parse decimals.",
}